File: practice/pepcoding/scala/L2/ArrayStrings/firstMissingPositiveNum.scala
import scala.compiletime.ops.boolean
object firstMissingPositiveNum{
    
    def firstMissingPositiveNum(nums: Array[Int]): Int= {
        val hasOne = nums.indices.foldLeft(false)((hasOne, index) => {
            if (nums(index) < 1) {
                nums(index) = 1
                hasOne
            } else hasOne || nums(index) == 1
      })

      if (!hasOne) 1
      else {
        //second pass go thru and make item at corresponding index negative
        nums.indices.foreach(index => {
          val indexValueToUpdate: Int = Math.abs(nums(index))

          if (indexValueToUpdate - 1 < nums.length) {
            nums(indexValueToUpdate - 1) = -Math.abs(nums(indexValueToUpdate - 1))
          }
        })

        nums.indices.find(index => nums(index) > 0).map(_ + 1).getOrElse(nums.length + 1)
      }
    }
    }
    def firstMissingPositive(nums: Array[Int]): Int = {
        var one:Boolean = false
        var n: Int = nums.size
        
        for(i <- 0 to n-1){
            
            if(nums(i) == 1) one = true
            
            if(nums(i) < 1 || n < nums(i)) {
                nums(i) = 1
            }
        }
        
        if(one == false) 1
        else{
            var val = 0
            for(i <- 0 to n-1){
                var val = nums(i).abs
                var idx = val-1
                nums(idx) = -nums(idx).abs
                
            }
        
            for(int i=0; i<n; i++){
                if(nums(i) > 0){
                    i+1;
                }
            }
            n+1
        }    
    }

}

File: practice/pepcoding/java/L2/ArrayStrings/firstMissingPositiveNum.java
public class firstMissingPositive{

	public static void main(String[] args){
		int[] arr = {3,4,-1,1};
		firstMissingPositive(arr);
	}

	public static int firstMissingPositive(int[] arr) {
		boolean one = false;
        int n = nums.length;
        
        for(int i=0; i<n; i++){
            
            if(nums[i] == 1) one = true;
            
            if(nums[i] < 1 || n < nums[i]) nums[i] = 1;
        }
        
        if(one == false) return 1;
        
        for(int i=0; i<n; i++){
            
            int val = Math.abs(nums[i]);
            int idx = val-1;
            nums[idx] = -Math.abs(nums[idx]);
            
        }
        
        for(int i=0; i<n; i++){
            if(nums[i] > 0){
                return i+1;
            }
        }
        return n+1;
	}

	public static int firstMissingPositiveII(int[] arr) {
		HashMap<Integer, Integer> hm = new HashMap<>();
		
		for(int i=0; i<arr.length; i++){
			hm.put(arr[i],1);
		}
		
		for(int i=1; i<arr.length; i++){
			
			if(!(hm.containsKey(i))){
				return i;
			}
		}
		return arr.length + 1;
	}

}
